extended/src/main/java/apoc/ml/OpenAI.java

```diff
@@ -0,0 +1,114 @@
+package apoc.ml;
+
+import apoc.Extended;
+import apoc.util.JsonUtil;
+import com.fasterxml.jackson.core.JsonProcessingException;
+import org.neo4j.procedure.Description;
+import org.neo4j.procedure.Name;
+import org.neo4j.procedure.Procedure;
+
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.List;
+import java.util.stream.Stream;
+
+import apoc.result.MapResult;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+
+
+@Extended
+public class OpenAI {
+
+    public static final String APOC_ML_OPENAI_URL = "apoc.ml.openai.url";
+
+    public static class EmbeddingResult {
+        public final long index;
+        public final String text;
+        public final List<Double> embedding;
+
+        public EmbeddingResult(long index, String text, List<Double> embedding) {
+            this.index = index;
+            this.text = text;
+            this.embedding = embedding;
+        }
+    }
+
+    private static Stream<Object> executeRequest(String apiKey, Map<String, Object> configuration, String path, String model, String key, Object inputs, String jsonPath) throws JsonProcessingException, MalformedURLException {
+        if (apiKey == null || apiKey.isBlank())
+            throw new IllegalArgumentException("API Key must not be empty");
+        String endpoint = System.getProperty(APOC_ML_OPENAI_URL,"https://api.openai.com/v1/");
+        Map<String, Object> headers = Map.of(
+                "Content-Type", "application/json",
+                "Authorization", "Bearer " + apiKey
+        );
+
+        var config = new HashMap<>(configuration);
+        config.putIfAbsent("model", model);
+        config.put(key, inputs);
+
+        String payload = new ObjectMapper().writeValueAsString(config);
+
+        var url = new URL(new URL(endpoint), path).toString();
+        return JsonUtil.loadJson(url, headers, payload, jsonPath, true, List.of());
+    }
+
+    @Procedure("apoc.ml.openai.embedding")
+    @Description("apoc.openai.embedding([texts], api_key, configuration) - returns the embeddings for a given text")
+    public Stream<EmbeddingResult> getEmbedding(@Name("texts") List<String> texts, @Name("api_key") String apiKey, @Name(value = "configuration", defaultValue = "{}") Map<String, Object> configuration) throws Exception {
+        // https://platform.openai.com/docs/api-reference/embeddings/create
+    /*
+    { "object": "list",
+      "data": [
+        {
+          "object": "embedding",
+          "embedding": [ 0.0023064255, -0.009327292, .... (1536 floats total for ada-002) -0.0028842222 ],
+          "index": 0
+        }
+      ],
+      "model": "text-embedding-ada-002",
+      "usage": { "prompt_tokens": 8, "total_tokens": 8 } }
+    */
+        Stream<Object> resultStream = executeRequest(apiKey, configuration, "embeddings", "text-embedding-ada-002", "input", texts, "$.data");
```

Comment:
@jexp @tomasonjo does make sense to make this and the other models configurable, via `configuration`?

Reply:
not sure exactly if I understand what you mean, but we could add model parameter under the configuration map. I would leave the prompts as a separate parameter.

Reply:
Yeah I mean something like this
```suggestion
        String model = configuration.getOrDefault("model", "text-embedding-ada-002")
        Stream<Object> resultStream = executeRequest(apiKey, configuration, "embeddings", model, "input", texts, "$.data");
```

Reply:
Not sure we need to do this on every call, as it's already done inside of executeRequest.